### PR TITLE
Add arguments to collect_results

### LINKED ISSUE #1901: How to select GPU programmatically in code
Hi,

I found this https://github.com/pytorch/examples/pull/73/files to set an env variable to choose GPU.
Is there a way to manually choose GPU programmatically in code? Thanks.

### LINKED ISSUE #1901: collect_results doesn't collect grads of non-parameter leaf inputs
### 🐛 Describe the bug

```
    out = gm(args)
    if only_fwd:
        return out
    if requires_bwd_pass(out):
        loss = reduce_to_scalar_loss(out)
        loss.backward()
    return collect_results(gm, out, None, [])
```

args can have gradients accumulated on it, but collect results ignores this entirely. These can matter if this graph further propagates changes to other parameters, and can impede minifier simplification.

### Error logs

_No response_

### Minified repro

_No response_

### PR DESCRIPTION
**Pull request by williamwen42**
Fixes https://github.com/pytorch/torchdynamo/issues/1901. Test script:
```python
import copy

import torch
import torch._dynamo as dynamo
import torch._dynamo.config

dynamo.config.repro_after = "dynamo"
dynamo.config.repro_level = 4

def custom_backend(gm: torch.fx.GraphModule, example_inputs):
    gm = copy.deepcopy(gm)
    for node in gm.graph.nodes:
        if len(node.args) > 1:
            node.target = torch.add
            node.args = (node.args[0], 0)
    gm.recompile()
    return gm

inp = torch.ones(5)
inp.requires_grad_(True)

@dynamo.optimize(custom_backend)
def foo(x):
    x = x * x
    return x.sum()

y = foo(inp)
print(y)
y.backward()
print(inp.grad)
```
Before, the script will finish but output an incorrect gradient. After the change, the accuracy minifier is triggered.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by ezyang** [comment]
Make sure to post full URL for issue reference, otherwise you are quoting a random pytorch issue

**Review by ezyang — APPROVED** [review]
Test? Or at the very least, how you verified it works?
